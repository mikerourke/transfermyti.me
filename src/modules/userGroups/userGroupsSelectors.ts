import { isNil } from "ramda";

import { activeWorkspaceIdSelector } from "~/modules/workspaces/workspacesSelectors";
import { createSelector } from "~/redux/reduxTools";
import type { ReduxState, UserGroup } from "~/typeDefs";

export const sourceUserGroupsSelector = createSelector(
  (state: ReduxState) => state.userGroups.source,
  (userGroupsById): UserGroup[] =>
    Object.values(userGroupsById).filter(({ name }) => !/Admin/gi.test(name)),
);

export const includedSourceUserGroupsSelector = createSelector(
  sourceUserGroupsSelector,
  (sourceUserGroups): UserGroup[] =>
    sourceUserGroups.filter((sourceUserGroup) => sourceUserGroup.isIncluded),
);

export const sourceUserGroupsForTransferSelector = createSelector(
  sourceUserGroupsSelector,
  (sourceUserGroups): UserGroup[] =>
    sourceUserGroups.filter((sourceUserGroup) =>
      isNil(sourceUserGroup.linkedId),
    ),
);

export const sourceUserGroupsInActiveWorkspaceSelector = createSelector(
  sourceUserGroupsSelector,
  activeWorkspaceIdSelector,
  (sourceUserGroups, workspaceId): UserGroup[] =>
    sourceUserGroups.filter(
      (userGroup) => userGroup.workspaceId === workspaceId,
    ),
);
