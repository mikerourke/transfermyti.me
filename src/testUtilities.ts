export { FAKES } from "~/redux/__fakes__/state";

type MockInstance = jest.MockInstance<AnyValid, AnyValid>;

interface WrapForMockingReturn {
  calls: AnyValid[];
  clear(): MockInstance;
  implementOnce(funcOrReturn: AnyValid): MockInstance;
  rejectAlways(err?: AnyValid): MockInstance;
  rejectOnce(err?: AnyValid): MockInstance;
  resolveMultiple(values: AnyValid[]): void;
  resolveOnce(value?: AnyValid): MockInstance;
  returnOnce(value?: AnyValid): MockInstance;
  returnMultiple(values: AnyValid[]): void;
  throwOnce(message?: string): MockInstance;
}

/**
 * Wrappers around someFunc.mock...() that Jest provides. Using
 * this avoids the annoying `someFunc as AnyValid` required to shut TypeScript
 * up.
 */
export function wrapForMocking(func: AnyValid): WrapForMockingReturn {
  return {
    calls: func?.mock?.calls,
    clear(): MockInstance {
      return func.mockClear();
    },
    implementOnce(funcOrReturn: AnyValid): MockInstance {
      if (typeof funcOrReturn === "function") {
        return func.mockImplementationOnce(funcOrReturn);
      } else {
        return func.mockImplementationOnce(() => funcOrReturn);
      }
    },
    rejectAlways(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValue(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    rejectOnce(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValueOnce(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    resolveMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockResolvedValueOnce(value);
      }
    },
    resolveOnce(value: AnyValid = null): MockInstance {
      return func.mockResolvedValueOnce(value);
    },
    returnOnce(value: AnyValid = null): MockInstance {
      return func.mockReturnValueOnce(value);
    },
    returnMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockReturnValueOnce(value);
      }
    },
    throwOnce(message: string = "error"): MockInstance {
      return func.mockImplementationOnce(() => {
        throw new Error(message);
      });
    },
  };
}

/**
 * Allows you to get around `delay()` calls in sagas.
 * @see https://github.com/jfairbank/redux-saga-test-plan/issues/257#issuecomment-503016758
 * @example
 *   return expectSaga(someSaga)
 *     .provide([
 *       { call: provideDelay },
 *     ])
 *     ...
 */
export function provideDelay(
  { fn }: AnyValid,
  next: AnyValid,
): null | VoidFunction {
  return fn.name === "delayP" ? null : next();
}
