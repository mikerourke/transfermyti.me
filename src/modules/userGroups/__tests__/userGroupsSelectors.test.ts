import cases from "jest-in-case";

import { FAKES } from "~/testUtilities";

import * as userGroupsSelectors from "../userGroupsSelectors";

describe("within userGroupsSelectors", () => {
  test("the sourceUserGroupsSelector returns array of values from state.source", () => {
    const result = userGroupsSelectors.sourceUserGroupsSelector(FAKES.REDUX_STATE);

    expect(result).toEqual(Object.values(FAKES.REDUX_STATE.userGroups.source));
  });

  cases(
    "the selectors match their snapshots",
    (options) => {
      const result = options.selector(FAKES.REDUX_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceUserGroupsSelector",
        selector: userGroupsSelectors.includedSourceUserGroupsSelector,
      },
      {
        name: "for the sourceUserGroupsForTransferSelector",
        selector: userGroupsSelectors.sourceUserGroupsForTransferSelector,
      },
      {
        name: "for the sourceUserGroupsInActiveWorkspaceSelector",
        selector: userGroupsSelectors.sourceUserGroupsInActiveWorkspaceSelector,
      },
    ],
  );
});
