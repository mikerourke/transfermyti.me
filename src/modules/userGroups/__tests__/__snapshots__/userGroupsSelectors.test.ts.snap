// Vitest Snapshot v1

exports[`within userGroupsSelectors > the selectors match their snapshots > for the sourceUserGroupsInActiveWorkspaceSelector 1`] = `
[
  {
    "entryCount": 0,
    "id": "5002",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "userGroups",
    "name": "Test Group B",
    "userIds": [],
    "workspaceId": "1001",
  },
]
`;

exports[`within userGroupsSelectors > the selectors match their snapshots > for the sourceUserGroupsInActiveWorkspaceSelector 2`] = `
[
  {
    "entryCount": 0,
    "id": "5002",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "userGroups",
    "name": "Test Group B",
    "userIds": [],
    "workspaceId": "1001",
  },
]
`;

exports[`within userGroupsSelectors > the selectors match their snapshots > for the sourceUserGroupsInActiveWorkspaceSelector 3`] = `
[
  {
    "entryCount": 0,
    "id": "5001",
    "isIncluded": false,
    "linkedId": "clock-user-group-01",
    "memberOf": "userGroups",
    "name": "Test Group A",
    "userIds": [],
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "id": "5002",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "userGroups",
    "name": "Test Group B",
    "userIds": [],
    "workspaceId": "1001",
  },
]
`;
